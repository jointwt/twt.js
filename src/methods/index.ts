import getData from '../utils/getData';
import fetchParams from '../utils/fetchParams';

import {
  APIOptions,
  PingResponse,
  NewAccount,
  Account,
  PostParams,
  TimelineParams,
  UserOrFeed,
  UnknownResponse,
} from '../types.d';

export default {
  ping: (options: APIOptions) => getData({
    ...fetchParams(options.url, 'ping'),
  }) as Promise<PingResponse>,

  register: (account: NewAccount, options: APIOptions) => getData({
    ...fetchParams(options.url, 'register'),
    body: account,
  }) as Promise<UnknownResponse>,

  authenticate: (account: Account, options: APIOptions) => getData({
    ...fetchParams(options.url, 'auth'),
    body: account,
  }) as Promise<UnknownResponse>,

  post: (params: PostParams, options: APIOptions) => getData({
    ...fetchParams(options.url, 'post'),
    headers: {
      Token: options.token,
    } as Record<string, string>,
    body: params,
  }) as Promise<UnknownResponse>,

  getTimeline: (params: TimelineParams, options: APIOptions) => getData({
    ...fetchParams(options.url, 'timeline'),
    headers: {
      Token: options.token,
    } as Record<string, string>,
    body: params,
  }) as Promise<UnknownResponse>,

  getDiscover: (params: TimelineParams, options: APIOptions) => getData({
    ...fetchParams(options.url, 'discover'),
    headers: {
      Token: options.token,
    } as Record<string, string>,
    body: params,
  }) as Promise<UnknownResponse>,

  follow: (userOrFeed: UserOrFeed, options: APIOptions) => getData({
    ...fetchParams(options.url, 'follow'),
    headers: {
      Token: options.token,
    } as Record<string, string>,
    body: userOrFeed,
  }) as Promise<UnknownResponse>,
};
