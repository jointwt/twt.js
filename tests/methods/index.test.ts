import methods from '../../src/methods';

jest.mock('../../src/utils/getData', () => (params: unknown) => Promise.resolve({  
  status: 200,
  message: 'ok',
  data: {
    message: 'successful mocked response from getData',
    params,
  }
}));

const testOptions = {
  url: 'https://sample.test.url',
  token: 'sample_test_token',
}

test('all methods are defined', () => {
  expect(typeof methods.ping).toBe('function');
  expect(typeof methods.register).toBe('function');
  expect(typeof methods.authenticate).toBe('function');

});

test('ping calls getData', async () => {
  const data = await methods.ping(testOptions);

  expect(data).toMatchSnapshot();

});

test('register calls getData', async () => {
  const testAccount = {
    username: 'sample_login',
    password: 'sample_password',
    email: 'sample@test.email.tld',
  };

  const data = await methods.register(testAccount, testOptions);

  expect(data).toMatchSnapshot();

});

test('authenticate calls getData', async () => {
  const testAccount = {
    username: 'sample_login',
    password: 'sample_password',
  };

  const data = await methods.authenticate(testAccount, testOptions);

  expect(data).toMatchSnapshot();

});

test('post calls getData', async () => {
  const testPost = {
    text: 'sample_text',
    postas: 'sample_user',
  };

  const data = await methods.post(testPost, testOptions);

  expect(data).toMatchSnapshot();

});

test('getTimeline calls getData', async () => {
  const testParams = {
    page: 1,
  };

  const data = await methods.getTimeline(testParams, testOptions);

  expect(data).toMatchSnapshot();

});

test('getDiscover calls getData', async () => {
  const testParams = {
    page: 1,
  };

  const data = await methods.getDiscover(testParams, testOptions);

  expect(data).toMatchSnapshot();

});

test('follow calls getData', async () => {
  const testParams = {
    nick: 'sample_user',
    url: 'sample_feed_url',
  };

  const data = await methods.follow(testParams, testOptions);

  expect(data).toMatchSnapshot();

});
