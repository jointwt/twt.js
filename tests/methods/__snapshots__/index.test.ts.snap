// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate calls getData 1`] = `
Object {
  "data": Object {
    "message": "successful mocked response from getData",
    "params": Object {
      "body": Object {
        "password": "sample_password",
        "username": "sample_login",
      },
      "method": "POST",
      "url": "https://sample.test.url/api/v1/auth",
    },
  },
  "message": "ok",
  "status": 200,
}
`;

exports[`follow calls getData 1`] = `
Object {
  "data": Object {
    "message": "successful mocked response from getData",
    "params": Object {
      "body": Object {
        "nick": "sample_user",
        "url": "sample_feed_url",
      },
      "headers": Object {
        "Token": "sample_test_token",
      },
      "method": "POST",
      "url": "https://sample.test.url/api/v1/follow",
    },
  },
  "message": "ok",
  "status": 200,
}
`;

exports[`getDiscover calls getData 1`] = `
Object {
  "data": Object {
    "message": "successful mocked response from getData",
    "params": Object {
      "body": Object {
        "page": 1,
      },
      "headers": Object {
        "Token": "sample_test_token",
      },
      "method": "POST",
      "url": "https://sample.test.url/api/v1/discover",
    },
  },
  "message": "ok",
  "status": 200,
}
`;

exports[`getTimeline calls getData 1`] = `
Object {
  "data": Object {
    "message": "successful mocked response from getData",
    "params": Object {
      "body": Object {
        "page": 1,
      },
      "headers": Object {
        "Token": "sample_test_token",
      },
      "method": "POST",
      "url": "https://sample.test.url/api/v1/timeline",
    },
  },
  "message": "ok",
  "status": 200,
}
`;

exports[`ping calls getData 1`] = `
Object {
  "data": Object {
    "message": "successful mocked response from getData",
    "params": Object {
      "method": "GET",
      "url": "https://sample.test.url/api/v1/ping",
    },
  },
  "message": "ok",
  "status": 200,
}
`;

exports[`post calls getData 1`] = `
Object {
  "data": Object {
    "message": "successful mocked response from getData",
    "params": Object {
      "body": Object {
        "post_as": "sample_user",
        "text": "sample_text",
      },
      "headers": Object {
        "Token": "sample_test_token",
      },
      "method": "POST",
      "url": "https://sample.test.url/api/v1/post",
    },
  },
  "message": "ok",
  "status": 200,
}
`;

exports[`register calls getData 1`] = `
Object {
  "data": Object {
    "message": "successful mocked response from getData",
    "params": Object {
      "body": Object {
        "email": "sample@test.email.tld",
        "password": "sample_password",
        "username": "sample_login",
      },
      "method": "POST",
      "url": "https://sample.test.url/api/v1/register",
    },
  },
  "message": "ok",
  "status": 200,
}
`;
